directories.go

```diff
@@ -0,0 +1,134 @@
+}
+
+func getDir(dir *string, basePath string, methodName string) string {
+	lock.Lock()
+	defer lock.Unlock()
+	if *dir == "" {
+		initDir(dir, basePath, methodName)
+	}
+	return *dir
+}
+
+func initDir(dir *string, basePath string, methodName string) {
+	*dir = filepath.Join(basePath, strings.ToLower(AppName))
+	if _, err := os.Stat(*dir); os.IsNotExist(err) {
+		err := os.MkdirAll(*dir, os.ModeDir|0755)
+		if err != nil {
+			log.Fatalf("unable to create or open %s %s: %v", methodName, *dir, err)
+		}
+	}
+}
+
+func PrintDirectories() {
+	data := DataDir()
+	cfg := ConfigDir()
+	state := StateDir()
+
+	fmt.Printf("Config directory: \t%s\n", cfg)
+	if data != cfg || state != cfg {
+		fmt.Printf("Mailbox directory:\t%s\n", data)
+		fmt.Printf("Logs directory:   \t%s\n", state)
+	}
+}
+
+func MigrateLegacyDataDir() {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		log.Fatal(err)
+	}
+	legacyDataDir := filepath.Join(homeDir, ".wl2k")
+	f, err := os.Stat(legacyDataDir)
+	switch {
+	case os.IsNotExist(err):
+		debug.Printf("tried to migrate from %s but it doesn't exist; nothing to do", legacyDataDir)
+		return
+	case err != nil:
+		log.Fatal(err)
+	case !f.IsDir():
+		log.Printf("tried to migrate from %s but it's not a directory, that's weird; ignoring", legacyDataDir)
+		return
+	}
+
+	log.Printf("Migrating your Pat files from %s to new locations", legacyDataDir)
+	migrateFile("config.json", legacyDataDir, ConfigDir())
+	migrateFile("mailbox", legacyDataDir, DataDir())
+	migrateFile("Standard_Forms", legacyDataDir, DataDir())
+
+	matches, err := filepath.Glob(filepath.Join(legacyDataDir, "rmslist*.json"))
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, match := range matches {
+		_, f := filepath.Split(match)
+		migrateFile(f, legacyDataDir, DataDir())
+	}
+
+	debug.Printf("migration from %s finished, renaming it", legacyDataDir)
+	err = os.Rename(legacyDataDir, legacyDataDir+"-old")
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func migrateFile(fileName string, fromDir string, toDir string) {
+	// make sure the old file is there
+	fromFile := filepath.Join(fromDir, fileName)
+	_, err := os.Stat(fromFile)
+	if errors.Is(err, os.ErrNotExist) {
+		// no legacy file, nothing to do
+		debug.Printf("File %s doesn't exist, not migrating it", fromFile)
+		return
+	} else if err != nil {
+		log.Fatal(err)
+	}
+
+	// touch the new file to make sure it's not there, and we can write to it
+	toFile := filepath.Join(toDir, fileName)
+	toFH, err := os.OpenFile(toFile, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o666)
+	if errors.Is(err, os.ErrExist) {
+		// new file already exists, don't clobber it
+		debug.Printf("new file %s already exists; ignoring %s", toFile, fromFile)
+		if toFH != nil {
+			defer toFH.Close()
+		}
+		return
+	} else if err != nil {
+		log.Fatal(err)
+	}
+	toFH.Close()
+	os.Remove(toFile)
+
+	debug.Printf("Migrating %s from %s to %s", fileName, fromDir, toDir)
+	err = os.Rename(fromFile, toFile)
+	if err != nil {
+		log.Fatal(err)
+	}
```

Comment:
this would be `return os.Rename(fromFile, toFile)` if you where to let this function return an error. It little less verbose IMHO.

Reply:
Oh yeah, that's helpful :+1: